libbeat/registry/backend/memlog/diskstore.go

```diff
@@ -0,0 +1,724 @@
+	tmpPath, err := s.checkpointTmpFile(filepath.Join(s.home, "checkpoint"), tbl)
+	if err != nil {
+		return err
+	}
+
+	fileTxID := s.txid + 1
+	fileName := fmt.Sprintf("%v.json", fileTxID)
+	checkpointPath := filepath.Join(s.home, fileName)
+
+	if err := os.Rename(tmpPath, checkpointPath); err != nil {
+		return err
+	}
+	s.syncHome()
+
+	// clear transaction log once finished
+	s.checkpointClearLog()
+
+	// finish current on-disk transaction by increasing the txid
+	s.txid++
+
+	s.dataFiles = append(s.dataFiles, dataFileInfo{
+		path: checkpointPath,
+		txid: fileTxID,
+	})
+
+	// delete old transaction files
+	if err := s.updateActiveSymLink(); err == nil {
+		s.removeOldDataFiles()
+	}
+
+	s.syncHome()
+	return nil
+}
+
+func (s *diskStore) updateActiveSymLink() error {
+	activeLink := filepath.Join(s.home, "active.json")
+
+	active, _ := activeDataFile(s.dataFiles)
+	if active == "" {
+		os.Remove(activeLink) // try, remove active symlink if present.
+		return nil
+	}
+
+	active = filepath.Base(active)
+	tmpLink := filepath.Join(s.home, "active.json.tmp")
+	if err := os.Symlink(active, tmpLink); err != nil {
+		return err
+	}
+
+	err := os.Rename(tmpLink, activeLink)
+	if err != nil {
+		return err
+	}
+
+	s.syncHome()
+	return nil
+}
+
+func (s *diskStore) syncHome() {
+	trySyncPath(s.home)
+}
+
+func (s *diskStore) removeOldDataFiles() {
+	L := len(s.dataFiles)
+	if L <= 1 {
+		return
+	}
+
+	removable, keep := s.dataFiles[:L-1], s.dataFiles[L-1:]
+	for i := range removable {
+		path := removable[i].path
+		err := os.Remove(path)
+		if err == nil || os.IsNotExist(err) {
+			continue
+		}
+
+		// ohoh... stop removing and construct new array of leftover data files
+		s.dataFiles = append(removable[i:], keep...)
+		return
+	}
+	s.dataFiles = keep
+}
+
+func (s *diskStore) checkpointTmpFile(baseName string, tbl *hashtable) (string, error) {
+	tempfile := baseName + ".new"
+	f, err := os.OpenFile(tempfile, os.O_RDWR|os.O_CREATE|os.O_TRUNC|os.O_SYNC, s.fileMode)
+	if err != nil {
+		return "", err
+	}
+
+	needsClose := true
+	defer func() {
+		if !needsClose {
+			f.Close()
+		}
+	}()
+
+	writer := bufio.NewWriterSize(&ensureWriter{f}, s.bufferSize)
+	enc := newJSONEncoder(writer)
+	if _, err = writer.Write([]byte{'['}); err != nil {
+		return "", err
+	}
+
+	first := true
+	for _, bin := range tbl.bins {
+		for i := range bin {
+			entry := &bin[i]
+
+			prefix := []byte(",\n")
+			if first {
+				prefix = prefix[1:]
+				first = false
+			}
+			if _, err = writer.Write(prefix); err != nil {
+				return "", err
+			}
+
+			err = enc.Encode(storeEntry{
+				Key:    unsafeString(entry.key),
+				Fields: entry.value,
+			})
+			if err != nil {
+				return "", err
+			}
+		}
+	}
+
+	if _, err = writer.Write([]byte("\n]")); err != nil {
+		return "", err
+	}
+
+	if err = writer.Flush(); err != nil {
+		return "", err
+	}
+
+	if err = syncFile(f); err != nil {
+		return "", err
+	}
+
+	needsClose = false
+	if err = f.Close(); err != nil {
```

Comment:
I am not sure I understand this correctly. `needsClose` is `false` meaning from the deferred function, `f` is going to be closed again after `f.Close` in called in the next line. What do I miss?

Reply:
The condition in the defer statement is indeed wrong.